### PR TITLE
[IMP] speedup routing map

### PR DESCRIPTION
**Pull request by Xavier-Do**
The current target, 16.0 is a little arbitrary, to discuss.

The routing map is an important part of the cold loading of a page. This can be improved multiple way.

### Improve the totally cold routing map
Most of the time spent is in the `werkzeug.routing.Rule.add` method representing 70~80% of the total routing map generation time.
This will need some low level thinking and optimisations
Another part is the `odoo.http._generate_routing_rules`, not so massive but still visible, mainly the `getmembers`, `is_method_a_route`, `build_controllers`, ... Hard to say what is possible to do but a possible improvement would be to generate a registry of routes inside the route decorator.

### Improve muti webiste routing map
Between website, the main difference is the rewrites rules:
`rewrites = dict([(x.url_from, x) for x in request.env['website.rewrite'].sudo().search(domain)])`

A simple assumption is that, if this list is empty, we could use the `None` routing map. 
Even if this list is not empty, we could alter the `None` routing map to remove some rules, add some redirect
A ModelConverter is also added if website is installed.

### Improve multidb routing map
The mro dynamic composition would make this difficult but investigating a per module routing map and combining them
based on the installed module could be an important improvement for the saas infrastructure. 

## What is actually done in this pr
Even if there are more possible improvements, werkzeug 2.2 will change how the routing map is managed, going from regexes to a decision tree. The change proposed in this pr change should remain compatible in theory or at least not to difficult to adapt.

### Lazy _compile_builder: 
~600 ms -> ~200ms. 
Instead of generating the build method for ~1100 routes, will only create the method when accessing this page and matching the rule, or when accessing the sitemap or searching a page (only a small amount of routes are actually build in this scenario)

### Increase re _MAX_SIZE
One important part of the routing map is compiling the regex for routes. They actually are the same for all databases, all website, .... and it take around 100~150 ms. Last test increasing the _MAX_SIZE from 512 to 2200 allowed to speedup the 
second routing map from ~200 to ~70 ms. This improvement is also cross db, but 2200 is maybe to low to benefit from that
since the re._MAX_SIZE is used by all regexes. The proposed size here is 4096.
This imp should be visible either for a db with mutiple website, or for a worker serving multiple db, or even if the routing map was invalidated and regenared.

Note that this cache could be conceptually dedicated to routes and populated exclusively from the @route decorator, avoiding to add database specific rules to this cache, but this is a little hard to combine.

### CODE REVIEW

File: odoo/addons/base/models/ir_http.py [hunk 1]
```diff
@@ -79,6 +79,34 @@ class SignedIntConverter(NumberConverter):
     num_convert = int
 
 
+class FasterRule(werkzeug.routing.Rule):
+    """
+    _compile_builder is a major part of the routing map generation and rules
+    are actually not build so often.
+    This classe allows to make this call to _compile_builder lazy
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_http.py:86]:
```suggestion
    This class makes this call to _compile_builder lazy.
```

**Review by rco-odoo — APPROVED** [review]

File: odoo/addons/base/models/ir_http.py [hunk 2]
```diff
@@ -80,6 +80,11 @@ class SignedIntConverter(NumberConverter):
 
 
 class LazyCompiledBuilder():
+    def __init__(self, _compile_builder, append_unknown):
+        self.callable = ...
+        self._compile_builder = _compile_builder
+        self.append_unknown = append_unknown
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_http.py:86]: I would make all attributes private.

**Comment by rdeodoo** [comment]
Note that I just read the commit message, this might be obvious or off-topic.
I am just curious about:
> Instead of generating the build method for ~1100 routes, will only create the method when accessing this page and matching the rule, or when accessing the sitemap or searching a page (only a small amount of routes are actually build in this scenario)

1. In other words, does that mean that we are now improving the cold loading perfs at the cost of perfs in hot state when accessing a page? (So, instead of encapsulated only once until next invalidation or server restart, it will be spread when accessing a page, maybe hours/days later?)
Or maybe I understood it wrong and it was actually done in both cases anyway and this is just removing the useless usage during routing map generation?
2. If yes:
2.1. Is it only once per same page access (so could even be hour or day later when a page is first reached), or every time a same page is accessed?
2.2. Do you have metric about the perf loss on those page access?

**Review by rco-odoo — APPROVED** [review]

**Comment by Xavier-Do** [comment]
> 1. In other words, does that mean that we are now improving the cold loading perfs at the cost of perfs in hot state when accessing a page? (So, instead of encapsulated only once until next invalidation or server restart, it will be spread when accessing a page, maybe hours/days later?)

There will still be an hot state, but the cache will populate only on demand for what is needed than populating the whole cache at once on the first page loaded.

>    Or maybe I understood it wrong and it was actually done in both cases anyway and this is just removing the useless usage during routing map generation?

With this solution, some useless build method wont be generated anymore (technical routes, backend route) Only frontend route will be generated in fine and only when requested.

> 2. If yes:
>    2.1. Is it only once per same page access (so could even be hour or day later when a page is first reached), or every time a same page is accessed?

Not sure I understand your question but if you access /event/<event> for the first time, the build method will be added to the routing map and never computed again until the routing map is invalidated. If the page /contact is accessed for the first time three days latter, this build method will be computed at this time and also added to the routing map. But I don't really expect a routing map to live for such a long time.

2.2. Do you have metric about the perf loss on those page access?
It is hard to say but this is would be round 400ms/1100 per page on first access on average, less than 0.4 ms, only on the first access of a few page, but still gaining 400 ms on the first accessed page of a worker when generating the routing map. But for most pages I assume that it would be even faster than that, and only take time for more complex routes.

**Comment by Julien00859** [comment]
Looks like `/event/<model("event.event"):event>/booth` was build twice:

```
WARNING odoo.addons.base.models.ir_http: building /event/<model("event.event"):event>/booth in 0.00034 
INFO werkzeug: "GET /event/design-fair-los-angeles-1/booth HTTP/1.1" 200 - 78 0.027 0.093
```

```
WARNING odoo.addons.base.models.ir_http: building /event/<model("event.event"):event>/booth in 0.00039 
[...]
INFO werkzeug: "GET /sitemap.xml HTTP/1.1" 200 - 256 0.150 0.428
```

Is it normal, is the sitemap different?

**Pull request closed without merging**